### PR TITLE
Adding GradCAM and Guided Backpropagation

### PR DESCRIPTION
**Pull request by vivekmig**
This adds GradCAM and Guided Backpropagation to Captum, with necessary variants. Also includes all relevant documentation and tests. (reopening PR from new fork)

This also updates the return types to the new docstring format and also adds interpolate_method as an argument, defaulting to "nearest", which can be used for all dimensionalities (3, 4, or 5D). A comment has also been added explaining that the original paper uses bilinear interpolation for GuidedGradCAM, which can be provided as an interpolation mode option in the case of 4D tensors.

### CODE REVIEW

File: captum/attr/_core/guided_backprop.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+        # Keeps track whether original input is a tuple or not before
+        # converting it into a tuple.
+        is_inputs_tuple = isinstance(inputs, tuple)
+
+        inputs = format_input(inputs)
+        gradient_mask = apply_gradient_requirements(inputs)
+
+        # set hooks for overriding ReLU gradients
+        self.model.apply(self._register_hooks)
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/guided_backprop.py]: Good point, will add, thanks!

File: captum/attr/_utils/attribution.py [hunk 2]
```diff
@@ -287,6 +288,35 @@ def __init__(self, forward_func, layer, device_ids=None):
...
+            interpolate_mode (str):  Method for interpolation, which
+                        muat be a valid input interpolation mode for
+                        torch.nn.functional. These methods are
+                        "nearest", "area", "linear" (3D-only), "bilinear"
+                        (4D-only), "bicubic" (4D-only), "trilinear" (5D-only)
+                        based on the number of dimensions of the given layer
+                        attribution.
+
+        Returns:
+            torch.Tensor:
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_utils/attribution.py]: We could use it, that's particularly useful when we want to use named returns, since the default structure is to just give a type and description. Here, the return is the upsampled attributions, so I just gave the type and description like the default, but if you prefer, can switch to naming this return!

File: captum/attr/_core/grad_cam.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            device_ids=self.device_ids,
+        )
+        summed_grads = torch.mean(
+            layer_gradients,
+            dim=tuple(x for x in range(2, len(layer_gradients.shape))),
+            keepdim=True,
+        )
+
+        non_neg_scaled_act = F.relu(
+            torch.sum(summed_grads * layer_eval, dim=1, keepdim=True)
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/grad_cam.py]: This is a good point! I agree it would be nicer to give the user the flexibility to look at both positive and negative values. The only issue is that the paper did seem to imply that the relu is part of their description of GradCAM, so I'll add a note in the docstring that taking the positive of the returned attrs is necessary to match the original algorithm.

File: captum/attr/_utils/attribution.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_utils/attribution.py]: I guess it's more of a convention whether we want to everywhere name our returns or not for public methods. If we have done that everywhere or in most cases it would be good to keep the convention.

File: captum/attr/_core/guided_backprop.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        return _format_attributions(is_inputs_tuple, gradients)
+
+    def _register_hooks(self, module):
+        if isinstance(module, torch.nn.ReLU):
+            hook = module.register_backward_hook(self._backward_hook)
+            self.backward_hooks.append(hook)
+
+    def _backward_hook(self, module, grad_input, grad_output):
+        if isinstance(grad_input, tuple):
+            return tuple(F.relu(inp) for inp in grad_input)
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/guided_backprop.py]: Yes, good point, it would be nice to support this as well! Will add in a separate PR as discussed.

File: captum/attr/_core/grad_cam.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                            to these arguments.
+                            Default: None
+
+            Returns:
+                *tensor* of **attributions**:
+                - **attributions** (*tensor*):
+                            Attributions based on GradCAM method.
+                            Attributions will be the same size as the
+                            output of the given layer, except for dimension 2,
+                            which will be 1 due to summing over channels.
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/grad_cam.py]: Hm, I think it is just dimension 2? For instance, if the input image is (1, 3, 32, 32) and the convolution output is (1, 20, 12, 12), the GradCAM attributions would be (1, 1, 12, 12). If interpolated after, the attributions would become (1, 1, 32, 32) and can be visualized in the input space. Let me know if I should reword this to make it clearer!

File: captum/attr/_core/guided_grad_cam.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        super().__init__(model)
+        self.grad_cam = LayerGradCam(model, layer, device_ids)
+        self.guided_backprop = GuidedBackprop(model)
+
+    def attribute(
+        self,
+        inputs,
+        target=None,
+        additional_forward_args=None,
+        chosen_input_index=0,
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/guided_grad_cam.py]: Sounds good! Will return all indices with None's replacing any which are not possible to upsample appropriately, as discussed.

File: captum/attr/_core/grad_cam.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/grad_cam.py]: I see, yeah, I was actually referring to this line: https://github.com/pytorch/captum/blob/e4445dbf1dac593c2f4644b109e974959baedd50/captum/attr/_core/grad_cam.py#L148

It computes the mean across all dimensions starting from 2 but when it multiplies by the eval which expands it. I think it is good to leave it as is.

File: captum/attr/_core/grad_cam.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python3
+import torch
+
+from .._utils.attribution import LayerAttribution
+from .._utils.common import format_input, _format_additional_forward_args
+from .._utils.gradient import compute_layer_gradients_and_eval
+
+
+class LayerGradCam(LayerAttribution):
+    def __init__(self, forward_func, layer, device_ids=None):
```
Comment (NarineK) [inline captum/attr/_core/grad_cam.py]: One thing I remembered is that this layer ones need probably GPU tests ...

Comment (vivekmig) [reply captum/attr/_core/grad_cam.py]: Good point, yeah it would be good to add some!
